File: packages/core/src/process.ts
import { spawn } from "node:child_process";
import { LocalConfigError } from "./errors.js";

export interface ProcessResult { stdout: string; stderr: string; exitCode: number }

export async function runProcess(command: string, args: string[], options: { cwd?: string; env?: NodeJS.ProcessEnv; allowFailure?: boolean } = {}): Promise<ProcessResult> {
  return await new Promise((resolve, reject) => {
    const child = spawn(command, args, {
      cwd: options.cwd,
      env: { ...process.env, ...options.env, GIT_TERMINAL_PROMPT: "0" },
      stdio: ["ignore", "pipe", "pipe"],
      windowsHide: true,
    });
    let stdout = "";
    let stderr = "";
    child.stdout.setEncoding("utf8").on("data", (chunk) => { stdout += chunk; });
    child.stderr.setEncoding("utf8").on("data", (chunk) => { stderr += chunk; });
    child.on("error", (error) => reject(new LocalConfigError(
      command === "git" ? "repository_failed" : "generic_error",
      `Cannot start ${command}: ${error.message}`,
      { command },
      { cause: error },
    )));
    child.on("close", (code) => {
      // Git porcelain uses leading columns for index/worktree state; trimming the
      // beginning would turn ` D path` into `D path` and corrupt path parsing.
      const result = { stdout: stdout.trimEnd(), stderr: stderr.trim(), exitCode: code ?? 1 };
      if (result.exitCode !== 0 && !options.allowFailure) {
        reject(new LocalConfigError("repository_failed", `${command} failed: ${result.stderr || result.stdout}`, { command, args, exitCode: result.exitCode }));
      } else resolve(result);
    });
  });
}

File: internal/core/process.go
package core

import (
	"bytes"
	"errors"
	"os"
	"os/exec"
	"strings"
)

type ProcessResult struct {
	Stdout   string
	Stderr   string
	ExitCode int
}

func RunProcess(command string, args []string, cwd string, extraEnv map[string]string, allowFailure bool) (ProcessResult, error) {
	cmd := exec.Command(command, args...)
	cmd.Dir = cwd
	env := append([]string{}, os.Environ()...)
	env = append(env, "GIT_TERMINAL_PROMPT=0")
	for key, value := range extraEnv {
		env = append(env, key+"="+value)
	}
	cmd.Env = env
	var stdout, stderr bytes.Buffer
	cmd.Stdout, cmd.Stderr = &stdout, &stderr
	err := cmd.Run()
	result := ProcessResult{Stdout: strings.TrimRight(stdout.String(), "\r\n"), Stderr: strings.TrimSpace(stderr.String())}
	if err == nil {
		return result, nil
	}
	var exitErr *exec.ExitError
	if errors.As(err, &exitErr) {
		result.ExitCode = exitErr.ExitCode()
		if allowFailure {
			return result, nil
		}
		message := result.Stderr
		if message == "" {
			message = result.Stdout
		}
		return result, NewError(ErrRepositoryFailed, command+" failed: "+message, map[string]any{"command": command, "args": args, "exitCode": result.ExitCode})
	}
	code := ErrGeneric
	if command == "git" {
		code = ErrRepositoryFailed
	}
	return result, WrapError(code, "Cannot start "+command+": "+err.Error(), err, map[string]any{"command": command})
}
